NppExec_parse_help_topics.py

```diff
@@ -0,0 +1,71 @@
+import sys
+import re
+import json
+import urllib.parse
+from html.parser import HTMLParser
+from pathlib import Path, PurePath
+
+class NppExecDocParser(HTMLParser):
+    CHARSET = 'cp1252'
+
+    def __init__(self):
+        super().__init__()
+        self.topics = {}
+        self.doc_title = ''
+        self.is_doc_title, self.is_content = False, False
+
+    def handle_starttag(self, tag: str, attrs: str):
+        _tag = tag.lower()
+        self.is_doc_title = (_tag == 'title')
+        self.is_content = \
+            re.match(r'h[1-6]', _tag) is not None or \
+            _tag in ['blockquote', 'body', 'div', 'p', 'pre']
+
+    def handle_data(self, data: str):
+        if self.is_doc_title:
+            self.doc_title = data.strip()
+            self.topics[self.doc_title] = { 'keywords': [] }
+            self.is_doc_title = False
+        elif self.is_content and bool(data.strip()):
+            try:
+                words = data.strip().split()
+                self.topics[self.doc_title]['keywords'].extend([w.replace("\"",'').replace("\'",'') for w in words])
```

Comment:
let's keep only unique words within a web page?

Reply:
I'm also thinking about removing meaningless words such as "a", "an", "the".

Reply:
One more idea: is it neccessary to create a list of words? Why not extract a complete text of each article (page) and then search in that text?

Reply:
> I'm also thinking about removing meaningless words such as "a", "an", "the".

Maybe the same pattern as a valid C++ identifier would be a good filter, excluding common non-words? 

The total page count is only ~60 right now, and most 3-letter search strings currently get ~20 matches. I think the main benefit of a stricter filter would be a smaller JSON dump, without much affect on user experience. But a filter is only convenient if you have an array of words, so the whole-text vs. words decision comes first.

For me, the best argument for whole-text would be matching phrases in context instead of isolated words. Whole-text is more suitable for regex, but compiling regex from user input will need more validation than a simple substring search.   

| whole-text |        | words      |       |
| :--        | :---   | :---       | :---  |
| **Pro**    |  preserves the space between words so you can match phrases           | **Pro**    |  easier to iterate over and probably faster (needs benchmarking) |
|            |  easily searchable with regex                                         |            |  exact matches are possible with `==`                       |
| **Contra** |  the JSON dump will have a sloppier structure                         | **Contra** |  can't match phrases or easily use regex      |
|            |  search space is wider and matches probably take longer (needs benchmarking)            |            |  non-words and punctuation need filtering out, e.g. `//` will match `//` anywhere, but not `// comment`; solution: only search for `comment`        |

Reply:
After some rethinking I came up:

#### [Option 1](https://github.com/rdipardo/nppexec/commit/dae49b53b34a69c29b5f24dd3896a9b469fbcf0f)

- a *string* of just the significant words, searched by substring
- 91.4 KB of formatted data

#### [Option 2](https://github.com/rdipardo/nppexec/commit/46d573f3ebd6dad59a593c1d342570c965cf7f10)

- a full text extract, excluding line breaks and non-words, dumped as a string and searched by regex
- 136 KB of formatted data

To be honest, I would say the full text option is just a poor substitute for `grep`ing the markup. Plain JavaSscript has no file system API that could read the content without actually serving it first. [AJAX requests](https://developer.mozilla.org/en-US/docs/Web/Guide/AJAX) might work, but recent browser versions don't allow it over the `file://` protocol, and a search box that only works on a server doesn't seen very useful . . .

Reply:
The .chm indexing uses words, so when you search for e.g. "nppexec script", it shows all occurrences of single "nppexec", single "script" and, correspondingly, combinations "nppexec script" and "script nppexec" (because both words are highlighted independently). 
If you type "exec", it does not find anything, though I would expect it to find "exec" in "nppexec", for example.

Reply:
> If you type "exec", it does not find anything, though I would expect it to find "exec" in "nppexec", for example.

Just to make sure we're looking at the same code:

- check out [the branch that implements "full-text" searching](https://github.com/rdipardo/nppexec/tree/wip/search-help-topics-by-content)
- edit `docs/search.js` around line 62 like so:

~~~js
Object.keys(topicIndex).forEach(function(k) {
  // ...
  var match = topicIndex[k]['content'].match(queryRE);
  if (match) {
    var words = match.input
    var lenMatch = match.index + query.length
    /* Matched start of input */
    if (match.index < words.indexOf(' ')) {
      console.log(`Found "${query}" in "${words.slice(match.index, lenMatch)}"`);
    /* Matched somewhere within input */
    } else  {
     console.log(`Found "${query}" in "${words.slice(words.slice(0, lenMatch).lastIndexOf(' ') + 1, lenMatch)}"`);
    }

    foundTopics.push({ 'topic': k, 'path': topicIndex[k]['path'] })
  }
});
~~~

- browse the docs with the console showing (Chrome, Edge: <kbd>Ctrl</kbd> + <kbd>SHIFT</kbd> + <kbd>I</kbd>; FireFox <kbd>F12</kbd>).
- type or paste in "exec"
- check the log for matches on "NppExec" or similar, e.g.,

<details>
<summary>'search.js:73 Found "exec" in "NppExec"'</summary>
<pre>
<code>
search.js:73 Found "e" in "NppE"
search.js:73 Found "e" in "NppE"
search.js:73 Found "e" in "NppE"
search.js:73 Found "e" in "NppE"
 Found "e" in "late"
 Found "e" in "NppE"
 Found "e" in "late"
 Found "e" in "se"
 Found "e" in "Unicode"
 Found "e" in "usage"
search.js:73 Found "e" in "Ge"
search.js:73 Found "e" in "some"
search.js:73 Found "e" in "some"
search.js:73 Found "e" in "NppE"
search.js:73 Found "e" in "NppE"
search.js:73 Found "e" in "E"
search.js:73 Found "e" in "NppE"
search.js:73 Found "e" in "E"
search.js:73 Found "e" in "Pre"
search.js:73 Found "e" in "Use"
search.js:73 Found "e" in "E"
search.js:73 Found "e" in "strle"
search.js:73 Found "e" in "e"
search.js:73 Found "e" in "Advance"
search.js:73 Found "e" in "compiling/inte"
search.js:73 Found "e" in "NppE"
search.js:73 Found "e" in "E"
search.js:73 Found "e" in "Run-time"
search.js:73 Found "e" in "cmd.e"
search.js:73 Found "e" in "Console"
search.js:73 Found "e" in "e"
search.js:73 Found "e" in "e"
search.js:73 Found "e" in "De"
search.js:73 Found "e" in "Clone"
search.js:73 Found "e" in "multiple"
search.js:73 Found "e" in "Note"
search.js:73 Found "e" in "Note"
search.js:73 Found "e" in "Trigge"
search.js:73 Found "e" in "course"
search.js:73 Found "e" in "Proce"
search.js:73 Found "e" in "give"
search.js:73 Found "e" in "date"
search.js:73 Found "e" in "Console"
search.js:73 Found "e" in "te"
search.js:73 Found "e" in "ke"
search.js:73 Found "e" in "Se"
search.js:73 Found "e" in "simple"
search.js:73 Found "e" in "compile"
search.js:73 Found "e" in "compile"
search.js:73 Found "e" in "source"
search.js:73 Found "e" in "Advance"
search.js:73 Found "e" in "doc\NppE"
search.js:73 Found "e" in "Collate"
search.js:73 Found "e" in "Console"
search.js:73 Found "e" in "Othe"
search.js:73 Found "e" in "use"
search.js:70 Found "e" in "e"
search.js:70 Found "e" in "e"
search.js:70 Found "e" in "E"
search.js:70 Found "e" in "e"
search.js:70 Found "e" in "E"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "Ex"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "cmd.ex"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "ex"
search.js:73 Found "ex" in "ex"
search.js:73 Found "ex" in "ex"
search.js:73 Found "ex" in "tex"
search.js:73 Found "ex" in "ex"
search.js:73 Found "ex" in "Nex"
search.js:73 Found "ex" in "tex"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "ex"
search.js:73 Found "ex" in "tex"
search.js:73 Found "ex" in "tex"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "SEL_SETTEX"
search.js:73 Found "ex" in "ex"
search.js:73 Found "ex" in "ex"
search.js:73 Found "ex" in "ex"
search.js:73 Found "ex" in "(ex"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "(cl.ex"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "doc\NppEx"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "ex"
search.js:73 Found "ex" in "NppEx"
search.js:73 Found "ex" in "tex"
search.js:73 Found "ex" in "NppEx"
search.js:70 Found "ex" in "Ex"
search.js:73 Found "ex" in "NppEx"
search.js:70 Found "ex" in "Ex"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "Exe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "cmd.exe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "tip.}}ProcessWaitClose("notepad++.exe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "exe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "exe"
search.js:73 Found "exe" in "Exe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "(cl.exe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "doc\NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "exe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:73 Found "exe" in "NppExe"
search.js:70 Found "exe" in "Exe"
search.js:73 Found "exe" in "NppExe"
search.js:70 Found "exe" in "Exe"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "Exec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "exec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "exec"
search.js:73 Found "exec" in "Exec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "doc\NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "exec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:73 Found "exec" in "NppExec"
search.js:70 Found "exec" in "Exec"
search.js:73 Found "exec" in "NppExec"
search.js:70 Found "exec" in "Exec"
</code>
</pre>
</details>

Reply:
Yes, you are right.
How are you opening the index.html in Firefox locally? If I just double-click it, then search for something and click the search result, a new Firefox window is opened. I assume the latter is the result of `Handle the DOMException` mentioned in the "search.js"?

Reply:
> If I just double-click it, then search for something and click the search result, a new Firefox window is opened.

The exception handler is doing that, here: https://github.com/d0vgan/nppexec/blob/7f07d03d77d7dc7cbc84f6a8599473729564215b/docs/search.js#L23

You can configure how documents are opened by removing "popup" from [the `windowFeatures` parameter][0]. It should only execute with browsing a `file://` path in FireFox. I figured a "modal" window was less annoying than a new page. At least it keeps the TOC in view.

As for highlighting the search string like in the CHM docs, we can do that by adding a script to each individual document and passing in the search string as a query parameter, then the document itself can use [the `window.find` API][1]. Injecting the script would best be done with batch processing.

That would be the only way, unfortunately. No browser I've tried will let a client-side script access the `.window` property of a frame. It's even surprising that FireFox is the only one to restrict a frame's `.location` property. If a future version of Chrome adopts the same policy, then every doc will have to be `window.open`'ed over the `file://` protocol, defeating the purpose of frames.

[0]: https://developer.mozilla.org/en-US/docs/Web/API/Window/open/#windowfeatures
[1]: https://developer.mozilla.org/en-US/docs/Web/API/Window/find

Reply:
> As for highlighting the search string like in the CHM docs, we can do that by adding a script to each individual document and passing in the search string as a query parameter [...]

I've [updated the branch] to give an idea of what that would look like. Just remember to run the Python script first.

[updated the branch]: https://github.com/rdipardo/nppexec/commit/060240f34b298ef29f9787511c49f7ee68a62692

Reply:
To me, it's better to remove the `popup` attribute since a new tab looks softer than a new window...
As for the latest `window.onload` function in each page, when I build a new .chm file with all these changes, each page produces the following error within the .chm :
```
Error in scenario on line 12 symbol 87, ')' is expected
```
The Microsoft's JScript parser does not seem to understand `u` in `/(%20|\+)/gu`. If I keep just `/(%20|\+)/g`, no error is shown.

Reply:
> If I keep just /(%20|\+)/g, no error is shown.

The 'u' flag is for preserving Unicode character boundaries, so it's not surprising the system doesn't support it. We don't need it when the docs are encoded in Windows-1252.

> Looks like it's also possible to highlight all occurrences as mentioned here:
https://stackoverflow.com/questions/9447488/match-all-occurrences-with-window-find

The (really old) `execCommand` API doesn't seem to have any effect in Chrome. I think maybe they just kept the name for backward compatability and ripped out the code.

The one thing I took from the S.O. post was the idea of looping over every match in the document. The [updated implementation](https://github.com/rdipardo/nppexec/commit/0463b31393a5902644b9cc75c124713b8fec9b76) uses [the `Range` API](https://developer.mozilla.org/en-US/docs/Web/API/range) and DOM manipulation to create virtual spans for every match. This will make the highlight style more consistent across browsers. Chrome's default selection background is gray, but in FireFox it's blue. (*FireFox now opens new tabs when frame access is blocked, as you can see in the screenshot.*)

<img width="600" alt="all-matches-hilite" src="https://github.com/d0vgan/nppexec/assets/59004801/1f7479c7-db5a-466f-adb1-a7befc0e6674">

`Range` may not be supported by the MS script host, so a feature test guards to whole `<script>` block.

Reply:
All right!
Just wondering: is it possible to extract the script into a single separate file to reference it in each document rather than embed it into each one?
Also, when I search for e.g. "calc", the _last_ occurrence of "calc" becomes active in e.g. "3.8.4. Calculations, strlen and so on". Is it possible to search in a reverse order or somehow remember the position of the very first occurrence to activate (scroll to) the first occurrence rather than the last one?

Reply:
Forgot to mention: the python script seems to stick together those words that were separated by a new line rather than by a space. For example, you can find the following within the generated "topics.js", I'm quoting:
```
process'esoutput
Thus,consoleapplication
NPP_RUNcommand
suchcommands
implementationof
```
and so on.

Reply:
> Is it possible to search in a reverse order or somehow remember the position of the very first occurrence to activate (scroll to) the first occurrence rather than the last one?

I added [a virtual class attribute to the matching ranges][0], so the topmost one can be looked up and scrolled into view. The script is now it's own file as requested. Setting up the doc tree is now done like this:

    python3 NppExec_parse_help_topics.py -a -u

To also format the JSON output, add, e.g., `-i 2 -s`

[0]: https://github.com/rdipardo/nppexec/commit/d0c61c13d553c2ad88bfb2738c32f32d8e3cde04#diff-905b35d661051eca3cf1afd44cbde959a1225129e077a74fbe8523ab66ed1ad8

Reply:
> the python script seems to stick together those words that were separated by a new line rather than by a space.

Should be fixed by https://github.com/rdipardo/nppexec/commit/38cd91062fed9142fe2f2d5ced6cb8e7c1a1f3cd

Reply:
Demo deployment with recent changes is online at <https://rdipardo.github.io/nppexec>

Reply:
Awesome!
The only thing that disturbs me - when I open it on my smartphone in Chrome or Opera, the topics are found, but it looks like they can't be shown.
Therefore maybe we need to get rid of frames and to use another approach? I mentioned the same idea here: https://github.com/d0vgan/nppexec/issues/84.

Reply:
> when I open it on my smartphone in Chrome or Opera, the topics are found, but it looks like they can't be shown.

FireFox for Android looks fine (at least in landscape orientation):

<img alt="Screenshot_20230830-120947_Firefox" width="580" src="https://github.com/d0vgan/nppexec/assets/59004801/b2129a1a-3376-4f43-94c3-f606a66e365e">

Tapping the `<input>` summons the virtual keyboard as expected, although it seems I have to tap the `RETURN` key before anything happens. The `<select>` gets populated, and tapping anywhere on the element opens a native radio-button list:

<div align="center">
  <img alt="Screenshot_20230830-121031_Firefox" width="340" src="https://github.com/d0vgan/nppexec/assets/59004801/da5d2427-2441-4461-8513-25e97558e73c">
</div>

I supect the issue with Chrome and downstream browsers comes from elements being styled badly on a small screen, since we haven't considered any responsive layout features.

At a minimum, the ["viewport" `<meta>` attribute][0] should be set for any document intended for mobile:

~~~html
<head>
<!-- ... -->
<meta name="viewport" content="width=device-width, initial-scale=1.0, shrink-to-fit=no">
<!-- ... -->
</head>
~~~~

There's also [CSS media queries][1] to adjust things according to screen dimension.

[0]: https://developer.mozilla.org/en-US/docs/Web/HTML/Viewport_meta_tag
[1]: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries

Reply:
So there's actually a couple things wrong with mobile Chrome, and none of them have to do with frames:

- tapping the virtual keyboard emits no `keyup` event unless the key is `RETURN`, `SPACEBAR`, or something non-alphanumeric like `"++"`. It's the same with mobile FireFox, except Chrome also doesn't fetch anything after the first event fires. It could be that repeated queries come from the autofill cache so the `<input>`s value is actually empty (the reason why some website forms only accept a typed-out e-mail address; pasting from the clipboard also bypasses `keyup` listeners);

- when `keyup` does fire, the `<select>` looks blank in collapsed mode if only two or fewer items are retrieved. I guess the minimum visible `"size"` attribute is greater than 2 in mobile Chrome . . . ?

The `<select>` also never expands, even with several items on it. That seems to be a [long-standing quirk][0] of mobile Chrome, but not FireFox, apparently.

I found workarounds for everything but the always-collapsed `<select>` box. Since desktop viewing is going to be the main use case, I'm not convinced it's a real issue when the native menu button list is the real control a mobile user will interact with. A mobile `<select>` is just a button that calls up the native UI control, even in FireFox.

With 860b327, the site is at least presentable on small screens, and the event handlers are more consistent. A `<select>`'s `change` event is emitted the same by a mouse click or a screen tap, so all pages now open more easily, even in the desktop version.

<div align="center">
<img alt="Screenshot_20230831-073646_Android" width="300" src="https://github.com/d0vgan/nppexec/assets/59004801/1873b5c3-90b3-4631-ab12-fe99765b515e">
</div>

<div align="center">
<img alt="Screenshot_20230831-073255_Android"  width="500" src="https://github.com/d0vgan/nppexec/assets/59004801/ac38187c-69c4-4738-a6c5-b0fef28761e6">
</div>


[0]: https://stackoverflow.com/q/19010505

Reply:
I've noticed a strange thing. The indexer does not add words which follow the `<a>` tag located after "See also:".
For example, "4.4.html" contains:
```
<p>See also: Do something (how-to) <a href="3.3.html">[3.3]</a>; Console output redirection <a href="4.5.html">[4.5]</a>; Grab date &amp; time to NppExec variable <a href="4.6.5.html">[4.6.5]</a>; Run and terminate PowerShell process <a href="4.6.7.html">[4.6.7]</a>; Send several commands to external tool <a href="4.6.9.html">[4.6.9]</a>.</p>
```
And the "topics.js" includes just this:
```
also: something (how-)
```
I'd expect it to include "Console output redirection", "Grab date time NppExec variable", "Run terminate PowerShell process", "Send several commands external tool" as well.
Is the indexer confused by `<a>` embedded into `<p>` ?

Reply:
> Is the indexer confused by `<a>` embedded into `<p>` ?

Apparently so: <https://github.com/d0vgan/nppexec/commit/0d54b669cc4a093ba2af5367d67ee0c0c39b691a>

Reply:
I realized I'd like to keep "if" in a text since it is a keyword and it is also used in e.g. "if-then-else" combinations.
As far as I can see, this can be achieved by removing `f` from `i[fnt]` in the `is_non_word`. However it would be more elegant to keep `is_non_word` as is and add the "if" as an explicit exclusion. To achieve that, do we need to transform `is_non_word` into a function?

Reply:
> To achieve that, do we need to transform `is_non_word` into a function?

The answer is [re.finditer](https://docs.python.org/3/library/re.html?highlight=finditer#re.finditer): 8cc8f03

Looping over each word let's you pick out the ones you want to keep.

One small drawback is that we have to erase characters in place, without affecting the string's length, or the loop counter gets thrown off. There may be extra space left over before/after punctuation characters, since we only erase the word, not the word and boundaries, the way regex replacement could do.

Reply:
> > To achieve that, do we need to transform `is_non_word` into a function?
> 
> <del>The answer is [re.finditer](https://docs.python.org/3/library/re.html?highlight=finditer#re.finditer): [8cc8f03](https://github.com/d0vgan/nppexec/commit/8cc8f0347165948ac4a33f8f9accdc6e43bff387)</del>

Having looked at [the docs](https://docs.python.org/3/library/re.html?highlight=finditer#re.sub) again, I finally guessed what you had in mind: 89f4ccc